tidy3d/plugins/resonance/resonance.py

```diff
@@ -0,0 +1,329 @@
+"""Find resonances in time series data
+"""
+
+from typing import Tuple
+from functools import partial
+
+import numpy as np
+import scipy.linalg
+import xarray as xr
+
+from pydantic import Field, NonNegativeFloat, PositiveInt, validator
+
+from ...components.base import Tidy3dBaseModel
+from ...components.types import ArrayLike
+from ...components.data.data_array import ScalarFieldTimeDataArray
+from ...constants import HERTZ
+from ...log import log, SetupError, ValidationError
+
+# ResonanceData will be used internally
+class ResonanceData(Tidy3dBaseModel):
+    """Data class for storing objects computed while running the resonance finder."""
+
+    eigvals: ArrayLike[complex, 1] = Field(
+        ..., title="Eigenvalues", description="Resonance eigenvalues."
+    )
+    complex_amplitudes: ArrayLike[complex, 1] = Field(
+        None, title="Complex amplitudes", description="Complex resonance amplitudes"
+    )
+    errors: ArrayLike[float, 1] = Field(
+        None, title="Errors", description="Rough eigenvalue error estimate."
+    )
+
+
+class ResonanceFinder(Tidy3dBaseModel):
+    """Tool that extracts resonance information (frequency, decay rate,
+    amplitude, and phase) from a time series of the following form.
+
+    Note
+    ----
+    .. math::
+
+        f(t) = \\sum_k a_k e^{i \\phi_k} e^{-2 \\pi i f_k t - \\lambda_k t}
+
+    Note
+    ----
+    We use the algorithm described in
+
+    Vladimir A. Mandelshtam and Howard S. Taylor,
+    "Harmonic inversion of time signals and its applications,"
+    J. Chem. Phys. 107, 6756 (1997).
+
+    Example
+    -------
+    t = np.linspace(0, 10000, 10000)
+    f1 = 2*np.pi*0.1 - 1j*0.002
+    f2 = 2*np.pi*0.2 - 1j*0.0005
+    sig = 2*np.exp(-1j*f1*t) + 3*1j*np.exp(-1j*f2*t)
+    resdata = ResonanceFinder(freq_window=(0.05, 0.25)).run(sig, 1)
+    """
+
+    freq_window: Tuple[float, float] = Field(
+        ...,
+        title="Window ``[fmin,fmax]``",
+        description="Window ``[fmin,fmax]`` for the initial frequencies. "
+        "The resonance finder is initialized with an even grid of frequencies between "
+        "fmin and fmax. The resonance finder then iteratively optimizes and prunes these "
+        "frequencies. Note that frequencies outside this window may be returned.",
+        units=HERTZ,
+    )
+
+    num_freqs_bounds: Tuple[PositiveInt, PositiveInt] = Field(
+        (20, 300),
```

Comment:
let's define the `20` and `300` as constant values at the top of the file, ie line 18 ish.
```py
NUM_FREQ_BOUND_MIN = 20
NUM_FREQ_BOUND_MAX = 300
```

```py
    num_freqs_bounds: Tuple[PositiveInt, PositiveInt] = Field(
         (NUM_FREQ_BOUND_MIN, NUM_FREQ_BOUND_MAX),
```